### PR TITLE
Adding Webhooks documentation

### PR DESCRIPTION
**Pull request by kramantab**
Issue # 739
[Issue 739][(url](https://github.com/tableau/server-client-python/issues/739))

### CODE REVIEW

**Comment by kramantab** [comment]
Hey @bcantoni and @jacalata .  I added webhooks documentation.  Please take a look and let me know what you think.  In addition to the overall review, here are a few things I did differently, let me know if these are ok:
- I added a version statement in the intro
- I have "luid" in the description for "id".  
- in the examples,  I put the authentication in comments and added a URL to point to the login sample script.   If we can agree on the format for this, we can "snippettize" this so it can be reused in all examples and we don't have to write it every time.  

This is my first attempt at TSC doc - so look forward to any feedback on what I can do to improve and help make this review process easy for you, or anything else.

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,5 @@
 source 'https://rubygems.org'
 gem 'github-pages', group: :jekyll_plugins
 
+
+gem "webrick", "~> 1.7"
```
Comment (bcantoni) [inline Gemfile:5]: I _think_ we should revert this change, but I'm curious why it was added.

If it was for testing/previewing the page locally, we can already do this with Bundler: https://tableau.github.io/server-client-python/docs/dev-guide#running-docs-locally

File: docs/api-ref.md [hunk 2]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
 
-The project resources for Tableau are defined in the `WorkbookItem` class. The class corresponds to the workbook resources you can access using the Tableau REST API. The workbook methods are based upon the endpoints for projects in the REST API and operate on the `WorkbookItem` class.
+**Attributes**
 
+Name | Description
+:--- | :---
+`id` |  The identifier (*luid*) for the webhook. You need this value to query a specific Webhook  with the `get_by_id` method or to delete a Webhook with the  `delete` method.
+`name`  |  The name of the webhook. You must specify this when you create an instance of the `WebhookItem`.
+`url` | The destination URL for the webhook. The webhook destination URL must be https and have a valid certificate. You must specify this when you create an instance of the `WebhookItem`.
+`event` | The name of the Tableau event that triggers your webhook.This  is either `api-event-name` or `webhook-source-api-event-name`: one of this is required to create an instance of the `WebhookItem`.  <br> The event name must be one of the supported events listed in the Trigger Events table. The event and webhook-source use different name values for the same event.
```
Comment (bcantoni) [inline docs/api-ref.md:4726]: ...one of _these_ is required...

File: docs/api-ref.md [hunk 3]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
 
+**Example**
 
+```py
+    import tableauserverclient as TSC
 
+    # Create new Webhook_item
+
+    new_webhook = TSC.WebhookItem()
+```
```
Comment (bcantoni) [inline docs/api-ref.md:4737]: For code sample like this which are "fenced" by the 3 backticks, you don't need the additional indenting. I.e. just slide these code rows all the way to the left, then it'll look consistent.

File: docs/api-ref.md [hunk 4]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
+Returns the new webhook item. 
+
+**Example**
+```py
+# import tableauserverclient as TSC
+# server = TSC.Server('https://SERVER')
+# sign in . For authentication examples, see https://github.com/tableau/server-client-python/blob/master/samples/login.py
+
+---
+
```
Comment (bcantoni) [inline docs/api-ref.md:4780]: omit - not used inside a code sample

File: docs/api-ref.md [hunk 5]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
+ # create a webhook item
+ with server.auth.sign_in(tableau_auth):
+    new_webhook = TSC.WebhookItem()
+    new_webhook.name ='testWebhook'
+    new_webhook.event ="workbook-refresh-failed" # alternately, you can also use new_webhook.source="webhook-source-event-workbook-refresh-failed" 
+    new_webhook.url ="https://webhook.site/6e4c957d-dd40-422c-8fc6-7151afe7fc0b"
```
Comment (bcantoni) [inline docs/api-ref.md:4786]: for consistency, I would do single space on either side of the equals sign

File: docs/api-ref.md [hunk 6]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
+Error|Description
+:---|:---
+`Webhook ID undefined`| Raises an exception if a `webhook_id` is not provided. 
+
+**Example**
+```py
+# import tableauserverclient as TSC
+# server = TSC.Server('https://SERVER')
+# sign in . For authentication examples, see https://github.com/tableau/server-client-python/blob/master/samples/login.py
+---
```
Comment (bcantoni) [inline docs/api-ref.md:4823]: remove

File: docs/api-ref.md [hunk 7]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
+
+Returns a list of all `ProjectItem` objects and a `PagainationItem`. Use  these values to iterate through the results. 
+
+**Example**
+
+```py
+# import tableauserverclient as TSC
+# server = TSC.Server('https://SERVER')
+# sign in . For authentication examples, see https://github.com/tableau/server-client-python/blob/master/samples/login.py
+---
```
Comment (bcantoni) [inline docs/api-ref.md:4857]: remove

File: docs/api-ref.md [hunk 8]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
+# get a list of all the webhooks on a site
+with server.auth.sign_in(tableau_auth):
+   all_webhooks, pagination_item = server.webhooks.get()
+print("\nThere are {} webhooks on site: ".format(pagination_item.total_available))
```
Comment (bcantoni) [inline docs/api-ref.md:4861]: add one blank line before here

File: docs/api-ref.md [hunk 9]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
+:---|:---
+`Webhook ID undefined`| Raises an exception if a `webhook_id` is not provided. 
+
+
+**Example**
+```py
+# import tableauserverclient as TSC
+# server = TSC.Server('https://SERVER')
+# sign in . For authentication examples, see https://github.com/tableau/server-client-python/blob/master/samples/login.py
+---
```
Comment (bcantoni) [inline docs/api-ref.md:4894]: remove

File: docs/api-ref.md [hunk 10]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
...
+with server.auth.sign_in(tableau_auth):
+  webhook = server.webhooks.get_by_id ('7d60d364-b9f5-4a9c-8aa5-4bdaa38c5dd3')
+print (webhook.name, webhook.url)
```
Comment (bcantoni) [inline docs/api-ref.md:4897]: add one blank line before the print

**Review by bcantoni — CHANGES_REQUESTED** [review]
Overall this looks good! I made a few editing suggestions here. We'll also need Jac's review since she's more familiar with webhooks.

File: docs/api-ref.md [hunk 11]
```diff
@@ -4700,21 +4700,221 @@ See [ViewItem class](#viewitem-class)
 
 
 ---
+## Webhooks
+<br>
+<br>
+Using the Tableau Server Client (TSC), you can create a new webhook, get a list of all the Webhooks, get details about a specific webhook, or delete a webhook.
```
Comment (bcantoni) [inline docs/api-ref.md:4706]: Let's check for consistency on Webhooks vs webhooks. I think to match the rest of the page, we'd want to use the lower-case version unless talking about a specific class or something.

File: Gemfile (hunk 1 above)
Comment (kramantab) [reply Gemfile:5]: @bcantoni I inadvertently added this - is there an easy way to remove this file?

File: docs/api-ref.md (hunk 8 above)
Comment (kramantab) [reply docs/api-ref.md:4861]: @bcantoni  - just to make sure I understand, are you asking to add a blank line between the two print statements?

Comment (bcantoni) [reply docs/api-ref.md:4861]: right - basically adding a blank line after the indented code block, so for this snippet it would look like:

```
# get a list of all the webhooks on a site
with server.auth.sign_in(tableau_auth):
   all_webhooks, pagination_item = server.webhooks.get()

print("\nThere are {} webhooks on site: ".format(pagination_item.total_available))
```

File: Gemfile (hunk 1 above)
Comment (bcantoni) [reply Gemfile]: yep, perfect!

**Review by jacalata — APPROVED** [review]

File: docs/api-ref.md [hunk 12]
```diff
@@ -4700,19 +4700,220 @@ See [ViewItem class](#viewitem-class)
...
 
-The project resources for Tableau are defined in the `WorkbookItem` class. The class corresponds to the workbook resources you can access using the Tableau REST API. The workbook methods are based upon the endpoints for projects in the REST API and operate on the `WorkbookItem` class.
+**Attributes**
+
+Name | Description
+:--- | :---
+`id` |  The identifier (*luid*) for the webhook. You need this value to query a specific Webhook  with the `get_by_id` method or to delete a Webhook with the  `delete` method.
+`name`  |  The name of the webhook. You must specify this when you create an instance of the `WebhookItem`.
+`url` | The destination URL for the webhook. The webhook destination URL must be https and have a valid certificate. You must specify this when you create an instance of the `WebhookItem`.
+`event` | The name of the Tableau event that triggers your webhook.This  is either `api-event-name` or `webhook-source-api-event-name`: one of these is required to create an instance of the `WebhookItem`.  <br> The event name must be one of the supported events listed in the Trigger Events table. The event and webhook-source use different name values for the same event.
```
[the remark this answers is not in the document]
Comment (kramantab) [reply docs/api-ref.md:4726]: @jacalata should I leave out the webhook-source-api-event-name entirely or just recommend using the api-event-name?

File: docs/api-ref.md [hunk 13]
```diff
@@ -4700,21 +4700,222 @@ See [ViewItem class](#viewitem-class)
...
+**Attributes**
 
+Name | Description
+:--- | :---
+`id` |  The identifier (*luid*) for the webhook. You need this value to query a specific webhook  with the `get_by_id` method or to delete a webhook with the  `delete` method.
+`name`  |  The name of the webhook. You must specify this when you create an instance of the `WebhookItem`.
+`url` | The destination URL for the webhook. The webhook destination URL must be https and have a valid certificate. You must specify this when you create an instance of the `WebhookItem`.
+`event` | The name of the Tableau event that triggers your webhook.This  is either `api-event-name` or `webhook-source-api-event-name`: one of these is required to create an instance of the `WebhookItem`. We recommend using the `api-event-name`. <br> The event name must be one of the supported events listed in the [Trigger Events](https://help.tableau.com/current/developer/webhooks/en-us/docs/webhooks-events-payload.html) table.
```
Comment (kramantab) [inline docs/api-ref.md:4726]: @jacalata I added a recommendation here to use the short name.  Let me know if we should actually just remove the 'webhook-source-api-event-name'.

**Comment by kramantab** [comment]
@bcantoni I believe I have resolved all your comments, let me know if I missed anything.  @jacalata I have one follow up comment for the event short/long names.  Let me know what you think.

File: docs/api-ref.md (hunk 13 above)
Comment (jacalata) [reply docs/api-ref.md:4726]: No, it's still in the code and some people are already using it, so it's better to have it documented.

**Review by bcantoni — APPROVED** [review]
@kramantab everything looks good to me. The Gemfile got broken again to something else, so I just pushed to your branch to fix it back. If you do a `git pull` on your side, you'll have the same in your local copy.

**Comment by kramantab** [comment]
Squashing the commits and doing a merge.
